esp-bootloader-esp-idf/src/partitions.rs

```diff
@@ -0,0 +1,671 @@
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, strum::FromRepr)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[repr(u8)]
+pub enum PartitionTablePartitionSubType {
+    /// It is the primary partition table.
+    Primary = 0,
+    /// It is a temporary partition table partition used by the partition table
+    /// OTA update functionality for downloading a new image.
+    Ota     = 1,
+}
+
+impl TryFrom<u8> for PartitionTablePartitionSubType {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        PartitionTablePartitionSubType::from_repr(value).ok_or(Error::Invalid)
+    }
+}
+
+/// Read the partition table.
+///
+/// Pass an implementation of [embedded_storage::Storage] which can read from
+/// the whole flash and provide storage to read the partition table into.
+pub fn read_partition_table<'a>(
+    flash: &mut impl embedded_storage::Storage,
+    storage: &'a mut [u8],
+) -> Result<PartitionTable<'a>, Error> {
+    flash
+        .read(PARTITION_TABLE_OFFSET, storage)
+        .map_err(|_e| Error::StorageError)?;
+
+    PartitionTable::new(storage)
+}
+
+/// A flash region is a "view" into the partition.
+///
+/// It allows to read and write to the partition without the need to account for
+/// the partition offset.
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct FlashRegion<'a, F> {
+    raw: &'a PartitionEntry<'a>,
+    flash: &'a mut F,
+}
+
+impl<F> embedded_storage::Region for FlashRegion<'_, F>
+where
+    F: embedded_storage::ReadStorage,
+{
+    fn contains(&self, address: u32) -> bool {
+        address >= self.raw.offset() && address < self.raw.offset() + self.raw.len()
+    }
+}
+
+impl<F> embedded_storage::ReadStorage for FlashRegion<'_, F>
+where
+    F: embedded_storage::ReadStorage,
+{
+    type Error = Error;
+
+    fn read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Self::Error> {
+        let address = offset + self.raw.offset();
+
+        if !self.contains(address) {
+            return Err(Error::OutOfBounds);
+        }
+
+        if !self.contains(address + bytes.len() as u32) {
+            return Err(Error::OutOfBounds);
+        }
+
+        self.flash
+            .read(address, bytes)
+            .map_err(|_e| Error::StorageError)
+    }
+
+    fn capacity(&self) -> usize {
+        self.raw.len() as _
+    }
+}
+
+impl<F> embedded_storage::Storage for FlashRegion<'_, F>
+where
+    F: embedded_storage::Storage,
+{
+    fn write(&mut self, offset: u32, bytes: &[u8]) -> Result<(), Self::Error> {
+        let address = offset + self.raw.offset();
+
+        if self.raw.is_read_only() {
+            return Err(Error::WriteProtected);
+        }
+
+        if !self.contains(address) {
+            return Err(Error::OutOfBounds);
+        }
+
+        if !self.contains(address + bytes.len() as u32) {
+            return Err(Error::OutOfBounds);
+        }
+
+        self.flash
```

Comment:
This is a read-erase-write cycle, right? We should probably implement the NOR flash traits so that users can opt into something less damaging.